clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java

```diff
@@ -68,12 +68,13 @@ public RecordMetadata get() throws InterruptedException, ExecutionException {
     public RecordMetadata get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
         // Handle overflow.
         long now = System.currentTimeMillis();
-        long deadline = Long.MAX_VALUE - timeout < now ? Long.MAX_VALUE : now + timeout;
+        long timeoutMillis = unit.toMillis(timeout);
+        long deadline = Long.MAX_VALUE - timeoutMillis < now ? Long.MAX_VALUE : now + timeoutMillis;
         boolean occurred = this.result.await(timeout, unit);
         if (nextRecordMetadata != null)
             return nextRecordMetadata.get(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
         if (!occurred)
```

Comment:
Seems like we should check if it's occurred before calling `nextRecordMetadata.get`.

Reply:
True. Timeout error throwing is moved up a bit